### PR TITLE
Issue #3808: Check indentation of this() and super()

### PR DESCRIPTION
**Pull request by pbludov**
Issue #3808

A call to `this(...)` and `super(...)` was ignored by the indentation check due to the AST incompatibility
```
    |      |--SUPER_CTOR_CALL -> super [3:8]
    |      |  |--LPAREN -> ( [3:13]
```
versus
```
    |   --EXPR -> [7:8]
    |      |  --METHOD_CALL -> ( [7:13]
    |      |      |--IDENT -> method_name [7:8]
```
and missing bindings in the `BlockParentHandler` and the `HandlerFactory`.

### CODE REVIEW

**Comment by rnveach** [comment]
@pbludov wercker produced an exception.
````
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing /pipeline/source/contribution/checkstyle-tester/src/main/java/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/grammars/InputRegressionJavaClass2.java
Caused by: java.util.NoSuchElementException
    at java.util.TreeMap.key (TreeMap.java:1327)
    at java.util.TreeMap.firstKey (TreeMap.java:290)
    at com.puppycrawl.tools.checkstyle.checks.indentation.LineSet.firstLineCol (LineSet.java:54)
    at com.puppycrawl.tools.checkstyle.checks.indentation.MethodCallHandler.getIndentImpl (MethodCallHandler.java:69)
    at com.puppycrawl.tools.checkstyle.checks.indentation.AbstractExpressionHandler.getIndent (AbstractExpressionHandler.java:85)
    at com.puppycrawl.tools.checkstyle.checks.indentation.MethodCallHandler.checkIndentation (MethodCallHandler.java:189)
    at com.puppycrawl.tools.checkstyle.checks.indentation.IndentationCheck.visitToken (IndentationCheck.java:332)
    at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit (TreeWalker.java:390)
````

**Comment by pbludov** [comment]
@rnveach This is much complicated then i thought.
It's more complicated than I thought. I forgot about the awful syntax of calling the base constructor for an inner class.

**Comment by rnveach** [comment]
Indentation is one of our more complicated checks.
CI is passing now. Please provide us regression on all projects for indentation.

**Comment by romani** [comment]
please beware that pitest is not working on Indentation, it is better to prove manually that to damage for coverage happens.
We got new suppression model for pitest, so may be regenerate report for Indentation and enforce mutation coverage as is if list of suppressions is not huge (manageable and reasonable to keep in shippable.sh)

**Comment by pbludov** [comment]
https://pbludov.github.io/issue-3808/index.html

No much informative for now. Projects must be divided by the ident offset.
Currently there a lot of warnings like
```
 'ctor def' child has incorrect indentation level 2X, expected level should be X.
```
I'll do it again later.

**Comment by pbludov** [comment]
done. https://pbludov.github.io/issue-3808/index.html
Many projects does not have a consistent indent offset. Due to this, some warnings seem to be wrong, like this one: https://pbludov.github.io/issue-3808/Hbase/index.html#A4

**Comment by rnveach** [comment]
> We got new suppression model for pitest, so may be regenerate report for Indentation and enforce mutation coverage as is if list of suppressions is not huge (manageable and reasonable to keep in shippable.sh)

Are we doing this in this PR?
Either way, we should base it's values on the code before this PR to ensure new code doesn't add new surviving mutations.

**Comment by romani** [comment]
> Are we doing this in this PR?

Not a demand.
Just recheck if smb will try to generate report for it.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+/**                                                                           //indent:0 exp:0
+ * This test-input is intended to be checked using following configuration:   //indent:1 exp:1
+ *                                                                            //indent:1 exp:1
+ * basicOffset = 2                                                            //indent:1 exp:1
+ * braceAdjustment = 0                                                        //indent:1 exp:1
+ * lineWrappingIndentation = 4                                                //indent:1 exp:1
+ * tabWidth = 4                                                               //indent:1 exp:1
+ */                                                                           //indent:1 exp:1
+class InputIndentationCtorCall { //indent:0 exp:0
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java]: issue is "lambda as the first parameter in call base class ctor", here are lambda in user defined test case.
but here is no "->" in input file...
1) As good practice ... reported test case is copied "as is" to Inputs to make sure it is covered for sure.
2) if smth else if going to be fixed (not releated to referenced issue, it is better to create new issue.)

Comment (pbludov) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java]: This issue is not related to lambda. As Richard said, the problem is that the call to base constructor is simply ignored by the indentation check.
But you are right, the harder the tests, the more they will cover.

Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java]: @romani Issue title was user made and they didn't know the underlying issue.
I have corrected the issue title. Feel free to modify it if you need more.

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java]: :) you mentioned it in title then in description

>Only happends with 'super' and only if the lambda comes first.

You did mention unrelated stuff too much.
That is why it is good practice to use test case from user as they do not know details, and just report what whey see.

Please fix description if it second "typo".

**Comment by pbludov** [comment]
@romani Pitest report is here: https://pbludov.github.io/issue-3808/pit-reports/index.html
It says 100% for `MethodCallHandler`. A couple of test cases  have been added, including the lambda test.

**Comment by romani** [comment]
Report is not accessible, 404.

**Comment by pbludov** [comment]
Fixed. I've forgot to include the folder to the commit.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java (hunk 1 above)
Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java]: > You did mention unrelated stuff too much.

@romani If you are referring to me, I only changed the title, not the description of the issue.

Comment (pbludov) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java]: node. @romani are you waiting from me to improve anything else for this PR?

**Comment by pbludov** [comment]
done. The PR is rebased and updated.

**Pull request merged**